settings_org: Improve robustness of organization settings.

pragatiagrawal31 opened the pull request:
This moves the configuration of widget type from settings_org to instead
live in respective HTML templates, via `data-widget-setting-type` and we
also remove `get_subsection_property_types` and refactor function
`populate_data_for_request` accordingly.

Fixes: #11708.

pragatiagrawal31 commented:
@timabbott @shubhamdhama FYI.

timabbott commented on static/js/settings_org.js:
```diff
@@ -10,86 +10,6 @@ exports.reset = function () {
-    },
-    description: {
-        type: 'text',
-    },
```
Why does this not require adding a widget type to the `description` textarea?  I guess the answer is it does, since I tested and reloading the page after saving a description change shows we're not actually sending the descriptions to the server.

A nice way to verify the code in this change is to look at each change you have here, and line them up against the changes to the templates.  If you removed an item here that isn't a bool without a corresponding change to the template, that's probably a bug.  

(Also, you can test by after clicking "save changes", reloading the page and seeing if your changes go through.  And you should test each and every widget this way, including the "select" dropdowns).  Or alternatively look at the payloads in the "network" tab to confirm the right changes got sent.

As a more high-level note, whenever you're doing a refactor, it's important to brainstorm at least one approach like this that you can use to be 100% confident that your changes didn't accidentally break the feature through this kind of omission -- most bugs end up looking kinda like this.

pragatiagrawal31 replied:
Ahh, I possibly missed that part while testing features. Yeah, I'll be careful while testing so that it doesn't end up creating a bug. And will do like to go through the "network" tab to confirm my changes, Thanks for letting me know this.

timabbott commented on static/js/settings_org.js:
```diff
@@ -10,86 +10,6 @@ exports.reset = function () {
-        },
-        email_address_visibility: {
-            type: 'integer',
-        },
```
Reading this, I'm curious -- should we be setting `data-setting-widget-type="integer"` for all the `select` elements?

timabbott replied:
(Might be a good follow-up project, as opposed to something to do in this PR -- try to make something like `settings_checkbox.handlebars` for select widgets)

pragatiagrawal31 replied:
Yeah, I guess maybe we can do make a generic widget for our select elements, I'll analyze the patterns and will be following up regarding its implementation(And if we do that then I guess we can set `data-setting-widget-type="integer"` for all the `select` elements.)

pragatiagrawal31 commented:
I've updated the PR, changes include : 
- Fade out time has been changed from 500ms to `800 ms`.
- Added a small commit, to make node tests of `test_change_save_button_state` function in `settings_org` more readable.
- `Discard` button fades out along with ` Save changes` button(in `300ms` using `show_hide_element` function). 
I tried both methods, but personally felt this behaviour of `Discard` button to be more desirable, so currently added fading out in 300ms, whatever be the final consensus, let me know, I'll be updating the PR accordingly.
 
**GIF** showing fading of `Discard` and `Save changes` button in `300 ms`.

![Mar-10-2019 14-59-22](https://user-images.githubusercontent.com/40331304/54083095-618dd400-4345-11e9-926b-33fd7503eb1c.gif)

 **GIF** showing the behaviour, where `Discard` and `Save changes` disappear immediately.

![Mar-10-2019 15-28-23](https://user-images.githubusercontent.com/40331304/54083426-31483480-4349-11e9-886f-93eb899b39b6.gif)
@timabbott FYI.

pragatiagrawal31 commented:
Thanks, for the review @shubhamdhama .
@timabbott FYI.

timabbott commented:
This looks great, merged, thanks @pragatiagrawal31!  

This will make it a lot easier to make changes to these without risk of problems.

As a follow-up, having played with the new "Saving -> Saved" behavior a bunch while QAing this, I think we should make a change to that logic.  The `setTimeout()` before the call to `exports.change_save_button_state` in `save_organization_settings` was useful (for ensuring you get to see the loading spinner in dev).  But we don't want to delay moving to the "Saved" state in the event that there is significant network/save delay.  So I'm thinking what we want is something like this:

```
time_start = new Date();
        exports.change_save_button_state(save_btn_container, "saving");                                            
        channel.patch({                                                                                            
            url: "/json/realm",                                                                                    
            data: data,                                                                                            
            success: function (response_data) {                                                                    
                failed_alert_elem.hide();                                                                          
                setTimeout(max(0, time_start + 300ms - new Date())), function () {
                    exports.change_save_button_state(save_btn_container, "succeeded");                                 
                });
                if (success_continuation !== undefined) {                                                          
                    success_continuation(response_data);                                                           
                }                                                                                                  
            },                                                                                                     

```

With more care for units than what I have above.

timabbott commented:
Just wanted to add, this sort of robustness cleanup is the kind of thing that makes no visible changes but is super valuable for preventing future bugs, so thanks for doing this!
